pip/qsharp/telemetry_events.py

```diff
@@ -0,0 +1,87 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+from .telemetry import log_telemetry
+import math
+
+# TODO: This should be populated by the build in the main module
+QSHARP_VERSION = "1.9.0"
+
+# TODO: Log extra params like qubit count (buckets), qubit type for RE, etc.?
+
+default_props = {"qsharp.version": QSHARP_VERSION}
+
+# For metrics such as duration, we want to capture things like how many shots or qubits in
+# the additional properties. However properties shouldn't be 'continuous' values, as they
+# create new 'dimensions' on the backend, which is limited, thus we want to bucket these properties.
+
+# See some of the notes at: https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/metrics-custom-overview#design-limitations-and-considerations
+
+
+def get_shots_bucket(shots: int):
+    if shots <= 1:
+        return 1
+    elif shots >= 1000000:
+        # Limit the buckets upper bound
+        return 1000000
+    else:
+        # Bucket into nearest (rounded up) power of 10, e.g. 75 -> 100, 450 -> 1000, etc.
+        return 10 ** math.ceil(math.log10(shots))
+
+
+# gets the order of magnitude for the number of qubits
+def get_qubits_bucket(qubits: int):
+    if qubits <= 1:
+        return 1
+    elif qubits >= 1000000:
+        # Limit the buckets upper bound
+        return 1000000
+    else:
+        # Bucket into nearest (rounded up) power of 10, e.g. 75 -> 100, 450 -> 1000, etc.
+        return 10 ** math.ceil(math.log10(qubits))
```

Comment:
@sezna , I was going to bucket these differently to shots. Anything less than a 100 shots is trivial and it can legitimately go up to 1 million. For qubits in a simulation, there's a big difference between 1 or 2, 10, or 20 qubits in a simulation (and you get into the realm of not simulatable around 30 or 40). So going up in powers of 10 isn't really granular enough.